caffe2/ideep/operators/conv_transpose_op.cc

```diff
@@ -0,0 +1,241 @@
+              filter_dims_mkldnn,
+              filter.get_data_type(),
+              stride_,
+              pad_tl(),
+              pad_br());
+      const bool weights_changed =
+          (cached_weights_descriptor_ != filter.get_descriptor());
+      if (weights_changed)
+        cached_weights_descriptor_ = filter.dup_descriptor();
+
+      if (training_mode_ || (weights_changed && !training_mode_)) {
+        auto filter_in = filter;
+        // TODO:it will be remove when deconvolution in mkl-dnn support iohw format.
+        filter_in.set_public_format(ideep::format::iohw);
+        filter_.init(expected_descriptor);
+        ideep::reorder::compute(filter_in, filter_);
+      }
+
+      // TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+      if (need_type_zero_pad(filter_.get_mkldnn_memory_desc_t())) {
+        return fallback_.Run(0);
+      }
+    } else {
+      CAFFE_ENFORCE_EQ(
+        filter.get_dim(1), X.get_dim(1),
+        "filter number must be equal to input channel number");
+
+      // TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+      if (need_type_zero_pad(filter.get_mkldnn_memory_desc_t())) {
+        return fallback_.Run(0);
+      }
+
+      Y_dims = CalcOutputDims(X, filter.get_dim(0));
+    }
+
+    if (InputSize() > BIAS) {
+      const auto& bias = Input(BIAS);
+      CAFFE_ENFORCE_EQ(bias.ndims(), 1, "bias must be 1D tensor");
+      CAFFE_ENFORCE(
+          bias.get_dim(0) == pre_converted ? filter.get_dim(0) : filter.get_dim(1),
+          "bias dimension must be equal to output channel number");
+
+      ideep::convolution_transpose_forward::compute(
+          X, pre_converted ? filter : filter_, bias, Y_dims, *Y, stride_, pad_tl(), pad_br());
+    } else {
+      ideep::convolution_transpose_forward::compute(
+          X, pre_converted ? filter : filter_, Y_dims, *Y, stride_, pad_tl(), pad_br());
+    }
+    return true;
+  }
+
+ private:
+  INPUT_TAGS(INPUT, FILTER, BIAS);
+  OUTPUT_TAGS(OUTPUT);
+
+  const bool training_mode_;
+  ideep::tensor filter_;
+  ideep::tensor::descriptor cached_weights_descriptor_;
+  // TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+  FALLBACK_OP fallback_;
+
+};
+
+class IDEEPConvTransposeGradientOp final : public IDEEPConvTransposeUnpoolBase {
+ public:
+  USE_IDEEP_DEF_ALIASES();
+  USE_IDEEP_CONV_TRANSPOSE_UNPOOL_BASE_FUNCTIONS();
+  // TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+  using FALLBACK_OP = IDEEPFallbackOp<ConvTransposeGradientOp<float, CPUContext>>;
+
+  IDEEPConvTransposeGradientOp(const OperatorDef& operator_def, Workspace* ws)
+      : IDEEPConvTransposeUnpoolBase(operator_def, ws),
+        no_bias_(OperatorBase::GetSingleArgument<int>("no_bias", false)),
+        fallback_(operator_def, ws) {
+    OPERATOR_NEEDS_FEATURE(
+        pad_l() == pad_r() && pad_t() == pad_b(),
+        "Uneven padding not supported.");
+    CAFFE_ENFORCE(
```

Comment:
```suggestion
    CAFFE_ENFORCE_NE(
```

Reply:
@BIT-silence ,this condition is not OutputSize() != 3,it is !(no_bias_ && OutputSize() == 3),so it is ok in here by CAFFE_ENFORCE，do you think so?